### PR TITLE
Define marker attributes on a external assembly

### PR DESCRIPTION
**Pull request by skarllot**
## The issue or feature being addressed
- #122
- #150

## Details on the issue fix or feature implementation
- Create the assembly named Jab.Attributes to define marker attributes

### CODE REVIEW

**Comment by skarllot** [comment]
The package content changed to:
image

**Comment by pakrym** [comment]
Thank you for your contribution! One of the benefits of Jab, as I see it, is not needing any external dependencies. 

There are reasons to have attributes as a separate assembly (like supporting InternalsVisibleTo), but it should be an opt-in rather than the default. 

I was thinking about having `Jab.Attributes` as a separate package and the Attributes.cs injection, conditional on whether the package is already included.

**Comment by skarllot** [comment]
Done, the Jab no longer depends on Jab.Attributes.

The package content was changed to:

- 📂 analyzers/
  - 📂 dotnet/
    - 📂 roslyn3.11/
      - 📄 Jab.Roslyn3.dll
    - 📂 roslyn4.0/
      - 📄 Jab.dll
- 📂 build/
  - 📄 Jab.props
  - 📄 Jab.targets
- 📄 README.md

File: src/Jab/build/Jab.props [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿<Project>
+    <PropertyGroup>
+        <JabAttributesReferenced>false</JabAttributesReferenced>
```
Comment (pakrym) [inline src/Jab/build/Jab.props:3]: What is this property used for?

File: src/Jab/Attributes.cs [hunk 2]
```diff
@@ -1,5 +1,7 @@
 ﻿// <auto-generated/>
 
+#if !JAB_ATTRIBUTES_REFERENCED
```
Comment (pakrym) [inline src/Jab/Attributes.cs:3]: Do you mind linking both projects to the same Attributes.cs file with `#if`s that control type accessibility? 

Similar to how .NET runtime does [here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/NullableAttributes.cs#L9-L13)

Comment (skarllot) [reply src/Jab/Attributes.cs:3]: No problem. Do you have preference where the attributes file should be located?

File: src/Jab/build/Jab.props (hunk 1 above)
Comment (skarllot) [reply src/Jab/build/Jab.props:3]: It is used to control the compiler constant on `Jab.targets` file

```xml
    <Target Name="_JabReferences" DependsOnTargets="ResolvePackageAssets">
        <PropertyGroup>
            <JabAttributesReferenced
                Condition="@(PackageDependencies->WithMetadataValue('Identity', 'Jab.Attributes')->Count()) != 0">true</JabAttributesReferenced>
            <DefineConstants Condition="$(JabAttributesReferenced)">$(DefineConstants);JAB_ATTRIBUTES_REFERENCED</DefineConstants>
        </PropertyGroup>
    </Target>
```

File: src/Jab/ServiceProviderBuilder.cs [hunk 3]
```diff
@@ -123,7 +124,12 @@ internal class ServiceProviderBuilder
     public ServiceProviderBuilder(GeneratorContext context)
     {
         _context = context;
-        _knownTypes = new KnownTypes(context.Compilation, context.Compilation.Assembly);
+
+        var assemblySymbol =
```
Comment (pakrym) [inline src/Jab/ServiceProviderBuilder.cs:128]: I am worried that we have two ways of detecting Jab.Attributes being referenced. Is it possible to unify to only looking at Compilation.SourceModule.ReferencedAssemblySymbols?

File: src/Jab/build/Jab.props (hunk 1 above)
Comment (pakrym) [reply src/Jab/build/Jab.props:3]: Is it required to be pre-initialized in a props file?


Can we 

```
    <Target Name="_JabReferences" DependsOnTargets="ResolvePackageAssets">
        <PropertyGroup>
            <_JabAttributesReferenced
                Condition="@(PackageDependencies->WithMetadataValue('Identity', 'Jab.Attributes')->Count()) != 0">true</JabAttributesReferenced>
            <DefineConstants Condition="'$(_JabAttributesReferenced)' == 'True'">$(DefineConstants);JAB_ATTRIBUTES_REFERENCED</DefineConstants>
        </PropertyGroup>
    </Target>
```

File: src/Jab/ServiceProviderBuilder.cs (hunk 3 above)
Comment (skarllot) [reply src/Jab/ServiceProviderBuilder.cs:128]: It will need to change the parameters of `ProcessModule` method

Comment (pakrym) [reply src/Jab/ServiceProviderBuilder.cs:128]: Oh, I mean the fact that we have a JAB_ATTRIBUTES_REFERENCED variable that controls Attributes.cs injection but KnownTypes work differently.

Can we switch to using the same logic when adding Attributes.cs?

Comment (pakrym) [reply src/Jab/ServiceProviderBuilder.cs:128]: Just looked, I guess not. RegisterPostInitializationOutput doesn't have enough information it seems.

File: src/Jab/build/Jab.props (hunk 1 above)
Comment (skarllot) [reply src/Jab/build/Jab.props:3]: No, it's not. I'll change as suggested

File: src/Jab/ServiceProviderBuilder.cs (hunk 3 above)
Comment (skarllot) [reply src/Jab/ServiceProviderBuilder.cs:128]: I'm not sure if I understood

Comment (pakrym) [reply src/Jab/ServiceProviderBuilder.cs:128]: Ignore me, what you have is good :)

File: src/Jab/Attributes.cs (hunk 2 above)
Comment (skarllot) [reply src/Jab/Attributes.cs]: Done, kept the one on Jab project

File: src/Jab/build/Jab.props (hunk 1 above)
Comment (skarllot) [reply src/Jab/build/Jab.props]: Done

**Comment by skarllot** [comment]
All, done as suggested. Not sure about the naming of the compiler constants.

**Review by pakrym — APPROVED** [review]
Looks great, thank you!

**Pull request merged**